auto-backup-v1

Kevin-Schnaubelt-Jr opened the pull request:
OVERVIEW:
Added auto-backup
Users can choose to turn auto-backup on, then select a time interval (1-60 minutes).
Optional backup maximum can be enabled, which will cull to a set max backups (10-100)
Utilizes threading to do so.

ADDITIONS:
imported time and threading

6 global variables

auto_backup() function. Takes the user through setting up the auto_backup, then runs backup_timer() on it's own thread.

backup_timer() function. Takes the users chosen interval and loops while backup_active is True. Thread sleeps for 1 second every loop with time.sleep() and increments a counter. Saves once user's threshold is reached and resets that counter. Thread locking happens whenever backup_active is changed to ensure it can can't be changed by more than 1 thread at a time.

manage_backups() function. If the user has set_max_backups to True, manage_backups will gather all the saves starting with 'BackupSave' and pop off and delete backups from oldest to newest, until the total number of backups are the users max_backups - 1. Default max_backups is set to 100 and resets to 100, redudentatly, just in case. Range is from 10 to 100. (save files can reach ~10mb, which noted in the auto backup menu.)

CHANGES: (Made sure to change as little as possible)
promptEnter() now has an extra message that can be passed into it to display extra info or errors. Defaults to an empty string.

auto_save() is slightly retrofitted. It's been given a bool argument, defaulted to true, for displaying the prompt asking if the user if they want to backup their save. manage_backups() added to it. Changed how backup files are named to only be BackupSave + the random string, instead of adding the len of all the files within that directory, as it had all your saves their and was inaccurate and became incompatible with deletions. Added error handling if the random string somehow rolls the same as an existing file. 

run() now has a new option, option 6, which displays an fstring to a global variable, which states "Start Auto Backup" until you enable auto-backup. Then it displays "Stop Auto Backup [set to {global variable} minutes]"

terminate() now more softly closes the program, canceling threads, though completely closing is still safe.

MISC:
Tried my best to preserve the original authors UI style. backup_timer has a commented out line that changes the interval to 5 seconds to debugging. Tested it a bunch, including using it while I play the game, and it seems to work. Sometimes the program can crash when loading a backup, but it's rare.

Kevin-Schnaubelt-Jr committed:
auto-backup-v1

wiseLlama0 commented:
First of all, thank you for your contribution! @Kevin-Schnaubelt-Jr 

I have some comments / suggestions to go through, I will mark those spots in the code and we can discuss them.

wiseLlama0 commented on SaveManager.py:
```diff
@@ -421,6 +436,124 @@ def view_saves():
+                break
+        
+        elapsed_time += 1
+        if elapsed_time >= interval_seconds:
+            # print(f"Auto backup created at {datetime.datetime.now()}") # Debug print statement
+            auto_save(display_prompt=False)
+            elapsed_time = 0  # Reset elapsed time after performing a backup
```
I think we should only do backups if there have been modifications to the save file since the last backup. Otherwise, if you remain afk in the game, or otherwise somehow play for a long time without hitting new save points, or leave this program running while not playing the game, your backup folder will become full of the same 10 - 100 saves, and you may even fill up your hard drive if you turn the backup limit off.

We can check for changes in the save files by looking at the modification timestamp.

Kevin-Schnaubelt-Jr replied:
Definitely!

wiseLlama0 commented on SaveManager.py:
```diff
@@ -421,6 +436,124 @@ def view_saves():
+    interval_seconds = interval * 60   # Convert interval to seconds
+    # interval_seconds = 5 # debug interval
+
+    while backup_active:
+        time.sleep(1)  # Thread sleeps for 1 second intervals to avoid busy-waiting
+        with backup_lock:
+            if not backup_active:  # Check if timer should still be running
+                break
+        
+        elapsed_time += 1
+        if elapsed_time >= interval_seconds:
```
Lets use the Unix Epoch time with time.time() to get a more accurate timer. We can capture the current time when the timer starts/restarts and use this to determine the elapsed time.

wiseLlama0 replied:
Also, I don't think it is necessary to check every second. Not that it will have a huge impact on performance either way, but I think we could safely move this up to every 10 seconds or something without any downsides. Especially if you have the timer set to every 60 minutes, being off by a few seconds would not be noticeable and would greatly reduce the number of wasted instructions.

Kevin-Schnaubelt-Jr replied:
I agree!

Kevin-Schnaubelt-Jr commented:
Let me make those changes and push again.

wiseLlama0 commented:
@Kevin-Schnaubelt-Jr About the crashing on loading, could you please provide me with some reproduction steps if you can, or at least the program output when it crashes. I would like to solve this crash before we release the update if we can. 

At first glance, my immediate suspicion would be that load_save() and backup_timer() both call the auto_save() function, which is probably leading to some race conditions on a file being managed. My guess would be that auto_save() calls the manage_backups() function which removes the oldest backup file. There is definitely a scenario where two threads can identify the same file as the oldest backup and both try to delete it, possibly causing a crash.

wiseLlama0 commented on SaveManager.py:
```diff
@@ -421,6 +436,124 @@ def view_saves():
+            auto_save(display_prompt=False)
+            elapsed_time = 0  # Reset elapsed time after performing a backup
+
+def manage_backups():
```
I would try adding the mutex lock to this entire manage_backups() function. To me there appears to be multiple possible race conditions here, including possible crashes or culling more backup saves than required.

Kevin-Schnaubelt-Jr committed:
update

Kevin-Schnaubelt-Jr committed:
update2

Kevin-Schnaubelt-Jr commented:
Made another commit! (update 2)

The latest commit (Update 2) has two major things.

adds a get_latest_save_timestamp() function. This returns the time stamp for the most recent file in the save directory (trying to make it only grab the last backup but I keep breaking it lol. Kinda tired.)

Anyway, the global variable is none on start. When you start the auto backup from auto_backup, it initializes this value. If you get to auto_save from somewhere else, or its otherwise still none, it will initialize it as well.

The other thing added was time.time() for a more accurate timing with backups.

adding a lock to manage_backups seemed to lock the whole function now

Kevin-Schnaubelt-Jr commented:
Left a bunch of debugs in... one sec

Kevin-Schnaubelt-Jr committed:
update3

Kevin-Schnaubelt-Jr commented:
OK update 3 is working without debugs. Again, only issues are the ones i mentioned in my comment after update2's commit.

wiseLlama0 commented:
> adding a lock to manage_backups seemed to lock the whole function now

 Ah, yes. My bad I didn't read the code carefully enough. We are getting a deadlock because on line 561 of backup_timer() we are already enabling the lock before calling auto_save(). Might need to create a second mutex lock for this or adjust where we are using this other lock. I'll have to take a look at it closer later.

Kevin-Schnaubelt-Jr commented:
> > adding a lock to manage_backups seemed to lock the whole function now
> 
> Ah, yes. My bad I didn't read the code carefully enough. We are getting a deadlock because on line 561 of backup_timer() we are already enabling the lock before calling auto_save(). Might need to create a second mutex lock for this or adjust where we are using this other lock. I'll have to take a look at it closer later.

Got it! Let me have a look at it. I think I'll be able to have an update by tomorrow.

wiseLlama0 commented:
> > > adding a lock to manage_backups seemed to lock the whole function now
> > 
> > 
> > Ah, yes. My bad I didn't read the code carefully enough. We are getting a deadlock because on line 561 of backup_timer() we are already enabling the lock before calling auto_save(). Might need to create a second mutex lock for this or adjust where we are using this other lock. I'll have to take a look at it closer later.
> 
> Got it! Let me have a look at it. I think I'll be able to have an update by tomorrow.

No rush. I'll take a look later tonight or tomorrow as well and see if I can fix the other issue you were talking about.

Kevin-Schnaubelt-Jr committed:
update004

Kevin-Schnaubelt-Jr committed:
update005

Kevin-Schnaubelt-Jr commented:
OK update 5!

I fixed the locking issue. Some functions would lock, then go into other functions that would also lock, locking out the whole program. I changed all locks to only happen when trying to edit backup_active (and one time just before auto saving inside of the timer.)

I made `get_latest_save_timestamp` to check the date of the last save in that characters save directory, backup or not, as I think that should be fine?  it's checking the latest save in that characters directory with the current game save.

If you want to backup your save from another place, like maybe you're making a new character, if you select to backup your save and your game save is the same as the last file in the directory, it's already backed up. You'll get this special prompt    
```
 if current_save_timestamp == last_backup_timestamp:
        if (display_prompt):
            promptEnter("No changes detected since last backup. Backup aborted.")
```

Did a bunch of testing and everything seems to work flawlessly!

Let me know what else you think should be improved.

Kevin-Schnaubelt-Jr commented:
Been using it flawless all day today!

wiseLlama0 commented:
> Been using it flawless all day today!

@Kevin-Schnaubelt-Jr 

Great. Sorry I haven't had a chance to review it again. Will do so tonight. 

By the way, feel free to add me on discord. Username is brunch11. Should be easier than talking here / on the nexus forums.

Kevin-Schnaubelt-Jr commented:
> brunch11

Sent! I'm Grendel92 on Discord.

Kevin-Schnaubelt-Jr committed:
copy of commit 92e59c0

Kevin-Schnaubelt-Jr committed:
update6

wiseLlama0 commented:
Approved!
